UIE-153 Environments Storybook and notifications decouple

msilva-broad opened the pull request:
### Jira Ticket: https://broadworkbench.atlassian.net/browse/UIE-153

<!-- ### Dependencies --->
<!-- Include any dependent tickets and describe the relationship. Include any relevant Jira tickets. --->

## Summary of changes:
UIE-153 Environments Storybook and notifications decoupling

Builds on earlier error.ts work to decouple notifications dependency, enabling Storybook component mounting of Environments component in preparation for AoU code sharing.

Environments has 3 stories in Storybook, and leverages simplistic mock swap-ins to mimic stateful interactions like pause and delete runtime.

### What
- draft of Storybook Stories for Environments complex component
- Storybook actions usage to assist dev inspection of mocked data calls, etc. being hit, with args inspection in response to component interactions like Runtime delete.
- decoupled notifications to be provided by react context provider, from new notifications package.
- errors.ts made compatible with composable notifications depenency.
- examples of components/hooks within Environments scope using useNotifactionsFromContext hook.
- converted src/libs/notifications.js to ts and improved key types.
- improved unit test coverage for notifications.ts and cleanup.

### Why
- showcase complex components in Storybook.
- AoU code sharing and broad UX/Dev ergonomics.

### Testing strategy
- [ ] unit tests pass
- [ ] 3 Storybook stories for Environments component work and interact as expected.

### Storybook
https://www.chromatic.com/library?appId=65fc89c9335768720ff8605a&buildNumber=38

msilva-broad committed:
UIE-153 Environments Storybook
- draft of Storybook Stories for Environments complex component
- decoupled notificaitons to be provided by react context provider, from new notifications package.
- errors.ts reworked to not use curried functions and support composeablity with NotificationContract implementor.
- consumers of error.ts fixed
- examples of components/hooks within Environments scope using useNotifactionsFromContext hook.

msilva-broad committed:
UIE-153 Environments Storybook
- minor cleanup

msilva-broad committed:
UIE-153 Code Sharing - Environments Storybook prep
- rebased
- converted src/libs/notifications.js to ts and improved key types.
- improved unit test coverage for notifications.ts and cleanup.

msilva-broad committed:
UIE-153 Code Sharing - Environments Storybook prep
- rebased
- moved fallback notificationsProvider out of notifications.ts and into error.ts to avoid jest module mocking confusion on consumers.
- minor cleanup in PreferenceCheckbox code while I was in there figuring out that notificationsProvider related mechanics were causing unit test breakage.

msilva-broad commented on src/workspaces/common/state/useWorkspaces.ts:
```diff
@@ -26,6 +27,7 @@ const defaultFieldsArgs: FieldsArg = [
  */
 export const useWorkspaces = (fieldsArg?: FieldsArg, stringAttributeMaxLength?: number): UseWorkspacesResult => {
   const signal = useCancellation();
+  const { reportError } = withErrorReporter(useNotificationsFromContext());
```
because useWorkspaces hook is used by Environments component, it must switch to the context driven notifications so that external consumers of Environments (AoU) can provide their own notifications implementation through react context.

msilva-broad commented on src/profile/notification-settings/PreferenceCheckbox.tsx:
```diff
@@ -4,12 +4,8 @@ import { Checkbox } from 'src/components/common';
-  updateUserAttributes,
-} from 'src/profile/notification-settings/utils';
+
+import { notificationEnabled, NotificationType, updateNotificationPreferences, updateUserAttributes } from './utils';
```
normalizing to relative path for nearby file (assuming this was the intent of an earlier change)

msilva-broad commented on src/libs/notifications.ts:
```diff
@@ -118,18 +145,6 @@ const NotificationDisplay = ({ id }) => {
                 },
                 ['Details']
               ),
-            !!action &&
```
removing support for action mechanic since it does not appear to be used by any consumers.  This need appears to be served by passing in rich ReactNode content through optional NotificationsProps.message arg instead.

msilva-broad committed:
UIE-153 Code Sharing - Environments Storybook prep
- minor cleanup
- added unit test coverage to notifications package

msilva-broad committed:
UIE-153 Code Sharing - Environments Storybook prep
- fixed file spelling error.

msilva-broad committed:
UIE-153 Code Sharing - Environments Storybook prep
- fixed file spelling error - pt2.

msilva-broad committed:
UIE-153 Code Sharing - Environments Storybook prep
- added missing yarn cache files

msilva-broad committed:
UIE-153 Code Sharing - Environments Storybook prep
- prettier formatting fixes
- added jsx/tsx to lint config

nawatts commented on package.json:
```diff
@@ -8,8 +8,10 @@
     "@dnd-kit/sortable": "^7.0.2",
     "@dnd-kit/utilities": "^3.2.1",
     "@microlink/react-json-view": "^1.22.2",
+    "@storybook/addon-actions": "^8.0.0",
```
Storybook packages should be under `devDependencies`.

nawatts commented on packages/notifications/package.json:
```diff
@@ -0,0 +1,49 @@
+    "react": "18.2.0",
+    "react-dom": "18.2.0",
+    "typescript": "~5.1.6",
+    "vite": "^4.3.9"
```
For consistency with other packages.

```suggestion
    "vite": "^4.5.2"
```

nawatts commented on packages/notifications/package.json:
```diff
@@ -0,0 +1,49 @@
+    "lib/types/**/*.d.ts"
+  ],
+  "dependencies": {
+    "@terra-ui-packages/components": "*",
```
Since this package is going to be published, it should specify a compatible version of components. Otherwise, components could update out from under it.

```suggestion
    "@terra-ui-packages/components": "0.0.6",
```

msilva-broad replied:
though taking in patch version ticks should be ok, right?

or even minor version ticks...

"^0.0.6"

nawatts replied:
That's fine, but it's also the same thing. Caret is "do not modify the left most non-zero number", which currently for components is the patch version.
https://github.com/npm/node-semver?tab=readme-ov-file#caret-ranges-123-025-004

msilva-broad replied:
on a relate note, we should probably stop using 0.x.y versions to make semantic versioning fully "active" across our packages..

nawatts replied:
Yeah, we've been kicking the can down the road until there was a need to start thinking about it. With packages starting to depend on each other, I agree that time has arrived.

nawatts commented on packages/notifications/vite.config.js:
```diff
@@ -0,0 +1,12 @@
+import { viteConfig as defineBaseConfig } from '@terra-ui-packages/build-utils';
+import { defineConfig } from 'vite';
+
+export default defineConfig((args) => {
+  const baseConfig = defineBaseConfig(args)
+  return {
+    ...baseConfig,
+    plugins: [
+      ...baseConfig.plugins,
+    ]
+  }
+});
```
Since there's no package specific configuration here, this could be simplified to:

```suggestion
export { viteConfig as default } from '@terra-ui-packages/build-utils';
```

msilva-broad replied:
hmmm.... kinda like having the config with template as is, in that it is ready for obvious extension.

nawatts replied:
Fair enough. If I'm being nitpicky, I think I'd still suggest removing the object spread stuff and just returning `baseConfig`.

msilva-broad replied:
ok. meet ya in the middle... kept a one-liner to pass-thru the base config

`export default defineConfig((args) => defineBaseConfig(args));`

nawatts reviewed (changes requested):
Haven't reviewed the code yet, but requesting a few changes to package.json files.

Optionally, Vite config could also be simplified.

cahrens commented:
@msilva-broad could you rebase this page to pick up the Chromatic integration? I'd love to see what your story looks like!

msilva-broad committed:
UIE-153 Code Sharing - Environments Storybook prep
- rebase to pickup storybook publishing
- package.json fixes from draft PR feedback.

msilva-broad committed:
UIE-153 Code Sharing - Environments Storybook prep
- updated lock file

msilva-broad committed:
UIE-153 Code Sharing - Environments Storybook prep
- simplified vite.config.js

nawatts commented on packages/notifications/src/useNotifications.ts:
```diff
@@ -0,0 +1,36 @@
+
+export type NotificationType = 'error' | 'warn' | 'info' | 'success' | 'welcome';
+export interface NotificationOptions {
+  detail: string | unknown;
```
Nit: SonarCloud complains about this
https://sonarcloud.io/project/issues?resolved=false&pullRequest=4741&id=DataBiosphere_terra-ui&open=AY579IL_8UV9StxhzyAO

`string | unknown` is effectively the same as `unknown`.

```suggestion
  detail: unknown;
```

nawatts commented on packages/notifications/src/useNotifications.ts:
```diff
@@ -0,0 +1,36 @@
+import { createContext, createElement, PropsWithChildren, useContext } from 'react';
+
+export type NotificationType = 'error' | 'warn' | 'info' | 'success' | 'welcome';
```
Consider not promoting `welcome` to the package? It's used in one (niche) place, where I suspect it would be fine to use `info` instead.

```suggestion
export type NotificationType = 'error' | 'warn' | 'info' | 'success';
```

msilva-broad replied:
hmmm... I debated this myself, but "welcome" seems a common enough thing to want... and while the logical difference to info is minimal, it is a11y related with a much more helpful "welcome notification" aria label being used... so I am loathe to remove that seeing as how it is not a huge impact on the types to include it.

nawatts replied:
Fair enough.

Though I would claim that it's single use in Terra UI, for a [notification shown when redirected from the old Firecloud](https://github.com/DataBiosphere/terra-ui/blob/5aa6b29bc6b86b36f2742079f45fb6639ae3148b/src/components/FirecloudNotification.js), suggests that it's not something frequently needed.

nawatts commented on packages/notifications/src/useNotifications.ts:
```diff
@@ -0,0 +1,36 @@
+};
+
+/** Provides notifications to descendents via React Context. */
+export const NotificationsProvider = (props: NotificationsProviderProps) => {
```
Super nit: convention.

```suggestion
export const NotificationsProvider = (props: NotificationsProviderProps): ReactNode => {
```

msilva-broad replied:
ah, this was a copy-and-modify of ThemeProvider, just checked and it lacks the ReactNode return convention also.  I can add to both.

nawatts commented on src/analysis/Environments/DeleteAppModal.ts:
```diff
@@ -21,6 +22,7 @@ export interface DeleteAppModalProps {
 
 export const DeleteAppModal = (props: DeleteAppModalProps): ReactNode => {
   const { app, onDismiss, onSuccess, deleteProvider } = props;
+  const { withErrorReportingInModal } = withErrorReporter(useNotificationsFromContext());
```
Nit: I think it might read a bit more clearly/conventional if the hook is called in its own statement.

```suggestion
  const notifications = useNotificationsFromContext();
  const { withErrorReportingInModal } = withErrorReporter(notifications);
```

But given how frequently this pair of functions appears, it may also be worth encapsulating in a hook `useErrorReporting` or something like that.

msilva-broad replied:
hmmm... I think we were going for a one-liner when we last looked at the code on my local.... and I am a bit nervous of hiding "....FromContext" in a secondary hook like useErrorReporting.... 

As I have it, it's a one-liner and the context usage is obvious... expressive is good.

nawatts replied:
> I am a bit nervous of hiding "....FromContext" in a secondary hook like useErrorReporting....

Why? `...fromContext` hooks will fail loudly if they're misused (used outside of the required context provider). The wrapper could be called `useErrorReportingFromContext` or something like that if desired, keeping the context use obvious, and would reduce the repeated boilerplate.

But in any case, this was more a suggestion for future consideration vs a request for changes to this PR.

nawatts commented on src/libs/error.test.ts:
```diff
@@ -0,0 +1,135 @@
+    jest.spyOn(console, 'error').mockImplementation(() => {});
+    const { reportError } = withErrorReporter(mockNotifications);
+
+    let outerThrow: unknown | null = null;
+
+    // Act
+    try {
+      await reportError('Things went BOOM', new Error('BOOM!'));
+    } catch (e) {
+      outerThrow = e;
+    }
```
It looks like this could be simplified to:

```suggestion
    // Act
    await reportError('Things went BOOM', new Error('BOOM!'));
```

Similar comment for mosts tests in this module.

msilva-broad replied:
not sure I follow... tests cannot throw unhandled exceptions... (considered a test fail) ... hence the try/catch pattern.

nawatts replied:
Exactly, the test should fail if `reportError` throws an exception. That's what the try/catch and `expect(outerThrow).toEqual(null);` currently does, but it uses a lot more code to do it.

msilva-broad replied:
ah, I missed the toEqual(null) reversal... must have forgotten to go back and clean these up... good catch!

cahrens commented:
It turns out that you CAN point to a branch directly in the URL. Feel free to update the PR template with this (I assume that slashes have to be escaped in some way).

https://<branch>--65fc89c9335768720ff8605a.chromatic.com

msilva-broad committed:
UIE-153 Code Sharing - Environments Storybook prep
- PR feedback
- added branch-specific Storybook link to PR template

msilva-broad committed:
UIE-153 Code Sharing - Environments Storybook prep
- fixed type error

msilva-broad commented:
> It turns out that you CAN point to a branch directly in the URL. Feel free to update the PR template with this (I assume that slashes have to be escaped in some way).
> 
> `https://<branch>--65fc89c9335768720ff8605a.chromatic.com`
> 
> There's also `https://www.chromatic.com/library?appId=65fc89c9335768720ff8605a&branch=<branch>`

> It turns out that you CAN point to a branch directly in the URL. Feel free to update the PR template with this (I assume that slashes have to be escaped in some way).
> 
> `https://<branch>--65fc89c9335768720ff8605a.chromatic.com`
> 
> There's also `https://www.chromatic.com/library?appId=65fc89c9335768720ff8605a&branch=<branch>`

worked this into the PR

msilva-broad committed:
UIE-153 Code Sharing - Environments Storybook prep
- rebase with dev

msilva-broad committed:
UIE-153 Code Sharing - Environments Storybook prep
- added missing NotificationsProvider to app root Main component.
